### ISSUE TITLE
Support for partially variadic functions

### ISSUE THREAD

**Issue by stevelinton** [gap-system/gap#89]
This is a very old idea, but I realised this evening it may be a bit easier than I had previously assumed.

The idea is to allow function expressions 

function(a,b,c,arg).....end;

This would require at least three arguments and assign the first three to a b and c and then a list containing any remaining ones to arg.

This requires extending the parser to recognise this case, which is easy and signal the numbert of arguments as -1-<number of names arguments> (this is consistent with the current setup where arg is signaled as -1).

Then when we code such a thing we just have to install the right handlers -- DoFail0Args..DoFail2Args (in this case) and then DoWrap3Args...DoWrap6Args.  So far these all exist.

Then we would need a new handler DoPartiallyUnwrapAndExecFunc (or some nicer name) which would be installed as the XArgs handler and would know to expect a list, look at the aritty of the function to find that it had to copy the first three elements into the first three lvars and shift the remaining ones left to the start of the list (adjusting the length accordingly) and then assign the list to the fourth lvar. 

There would need to be some printing and other stuff, but it's a lot less work than I thought.

**Comment by stevelinton** [comment]
This turned out to be much easier than I expected. There is basically working solution in my fork in 
sl/partial-variadic. Some work is still needed to correct some error messages and suchlike as well as testing and documentation.

**Comment by alex-konovalov** [comment]
For info, after resolving #126, the test of the manual examples now works again, so I was able to check the impact PR #90 had them. Now it has the following diffs:
```
########> Diff in [ "./language.xml", 1754, 1777 ]
# Input is:
position := function ( list, obj, arg )    
    local pos;
    if 0 = Length(arg) then
      pos := 0;
    else
      pos := arg[1];
    fi;
    repeat
      pos := pos + 1;
      if pos > Length(list) then
        return fail;
      fi;
    until list[pos] = obj;
    return pos;
   end;
# Expected output:
function( list, obj, arg ) ... end
# But found:
Syntax warning: New syntax used -- intentional? in stream line 3
    if 0 = Length(arg) then
     ^
function( list, obj, arg ) ... end
########
########> Diff in [ "./../../lib/function.g", 198, 209 ]
# Input is:
NumberArgumentsFunction(function(a,arg) return 1; end);
# Expected output:
-2
# But found:
Syntax warning: New syntax used -- intentional? in stream line 1
NumberArgumentsFunction(function(a,arg) return 1; end);
                                             ^
-2
########
########> Diff in [ "./../../lib/function.g", 398, 405 ]
# Input is:
f:=ReturnTrue;  
# Expected output:
function( arg ) ... end
# But found:
function( args ) ... end
########
########> Diff in [ "./../../lib/function.g", 425, 432 ]
# Input is:
f:=ReturnFalse;  
# Expected output:
function( arg ) ... end
# But found:
function( args ) ... end
########
########> Diff in [ "./../../lib/function.g", 452, 459 ]
# Input is:
oops:=ReturnFail;  
# Expected output:
function( arg ) ... end
# But found:
function( args ) ... end
########
```
The warning is there temporary for a reason. I may convert these to examples to `Log` elements instead of `Example` to avoid dozens of test failure notifications emails. But the three other examples show `args` instead of `arg`, are they correct?

**Comment by stevelinton** [comment]
I've fixed the args and arg thing in commit 99fc5f3 on master.

**Comment by alex-konovalov** [comment]
Thanks, and I've adjusted the two manual examples in commit f4091d28f46926696d1f61fa3497d301b007ecf9 - seems like this issue may be closed.
